libraries/plugins/rc/rc_plugin.cpp

```diff
@@ -0,0 +1,373 @@
+            block_info.usage[i] = count.resource_count[i]*int64_t( params.resource_unit );
+
+            pool = pool - block_info.decay[i] + block_info.budget[i] - block_info.usage[i];
+         }
+         pool_obj.last_update = gpo.time;
+      } );
+
+   std::shared_ptr< exp_rc_data > export_data =
+      steem::plugins::block_data_export::find_export_data< exp_rc_data >( STEEM_RC_PLUGIN_NAME );
+   if( export_data )
+      export_data->block_info = block_info;
+}
+
+void rc_plugin_impl::on_first_block()
+{
+   std::string resource_params_json = steem::jsonball::get_resource_parameters();
+   fc::variant resource_params_var = fc::json::from_string( resource_params_json, fc::json::strict_parser );
+   std::vector< std::pair< fc::variant, std::pair< fc::variant_object, fc::variant_object > > > resource_params_pairs;
+   fc::from_variant( resource_params_var, resource_params_pairs );
+
+   _db.create< rc_resource_param_object >(
+      [&]( rc_resource_param_object& params_obj )
+      {
+         for( auto& kv : resource_params_pairs )
+         {
+            auto k = kv.first.as< rc_resource_types >();
+            fc::variant_object& vo = kv.second.first;
+            fc::mutable_variant_object mvo(vo);
+            mvo["time_unit"] = int8_t( vo["time_unit"].as< rc_time_unit_type >() );
+            fc::from_variant( fc::variant( mvo ), params_obj.resource_param_array[ k ] );
+         }
+
+         ilog( "Genesis params_obj is ${o}", ("o", params_obj) );
+      } );
+   _db.create< rc_pool_object >(
+      [&]( rc_pool_object& pool_obj )
+      {
+         for( size_t i=0; i<STEEM_NUM_RESOURCE_TYPES; i++ )
+            pool_obj.pool_array[i] = 0;
+         pool_obj.last_update = _db.get_dynamic_global_properties().time;
+
+         ilog( "Genesis pool_obj is ${o}", ("o", pool_obj) );
+      } );
+   return;
+}
+
+} // detail
+
+rc_plugin::rc_plugin() {}
+rc_plugin::~rc_plugin() {}
+
+void rc_plugin::set_program_options( options_description& cli, options_description& cfg ){}
+
+void rc_plugin::plugin_initialize( const boost::program_options::variables_map& options )
+{
+   ilog( "Initializing resource credit plugin" );
+
+   my = std::make_unique< detail::rc_plugin_impl >( *this );
+
+   try
+   {
+      block_data_export_plugin* export_plugin = appbase::app().find_plugin< block_data_export_plugin >();
+      if( export_plugin != nullptr )
+      {
+         ilog( "Registering RC export data factory" );
+         export_plugin->register_export_data_factory( STEEM_RC_PLUGIN_NAME,
+            []() -> std::shared_ptr< exportable_block_data > { return std::make_shared< exp_rc_data >(); } );
+      }
+
+      chain::database& db = appbase::app().get_plugin< steem::plugins::chain::chain_plugin >().db();
+
+      my->_post_apply_block_conn = db.add_post_apply_block_handler( [&]( const block_notification& note ){ my->on_post_apply_block( note ); }, *this, 0 );
```

Comment:
We need to have a conversation regarding whether RC based filtering should happen pre or post tx.

If we do it pre-tx then we can filter out a potentially costly computation. However, this requires refactoring the `get_or_create_rc_account_object` because we might create objects for accounts that don't exist in consensus. This creates the opposite scenario where RC calculations are the more costly computation.